src/transformers/models/molmo/modular_molmo.py

```diff
@@ -0,0 +1,2289 @@
+        # therefore the input hidden states gets silently casted in float32. Hence, we need
+        # cast them back in the correct dtype just to be sure everything works as expected.
+        # This might slowdown training & inference so it is recommended to not cast the LayerNorms
+        # in fp32.
+
+        input_dtype = query_states.dtype
+        if input_dtype == torch.float32:
+            if torch.is_autocast_enabled():
+                target_dtype = torch.get_autocast_gpu_dtype()
+            # Handle the case where the model is quantized
+            elif hasattr(self.config, "_pre_quantization_dtype"):
+                target_dtype = self.config._pre_quantization_dtype
+            else:
+                target_dtype = self.q_proj.weight.dtype
+
+            logger.warning_once(
+                f"The input hidden states seems to be silently casted in float32, this might be related to"
+                f" the fact you have upcasted embedding or layer norm layers in float32. We will cast back the input in"
+                f" {target_dtype}."
+            )
+
+            query_states = query_states.to(target_dtype)
+            key_states = key_states.to(target_dtype)
+            value_states = value_states.to(target_dtype)
+
+        attn_output = _flash_attention_forward(
+            query_states,
+            key_states,
+            value_states,
+            None,
+            q_len,
+            dropout=dropout_rate,
+            is_causal=False,
+            use_top_left_mask=self._flash_attn_uses_top_left_mask,
+        )
+
+        attn_output = attn_output.reshape(batch_size, q_len, self.embed_dim).contiguous()
+        attn_output = self.o_proj(attn_output)
+
+        if not output_attentions:
+            attn_weights = None
+
+        return attn_output, attn_weights
+
+
+MOLMO_POOLING_ATTENTION_CLASSES = {
+    "eager": MolmoPoolingAttention,
+    "sdpa": MolmoPoolingSdpaAttention,
+    "flash_attention_2": MolmoPoolingFlashAttention2,
+}
+
+
+@add_start_docstrings(
+    """The adapter model from MOLMO that takes in image hidden states from vision tower.""",
+    MOLMO_START_DOCSTRING,
+)
+class MolmoAdapterModel(MolmoPreTrainedModel):
+    config_class = MolmoPoolingConfig
+    main_input_name = "image_features"
+
+    def __init__(self, config: MolmoPoolingConfig):
+        super().__init__(config)
+
+        attention_class = MOLMO_POOLING_ATTENTION_CLASSES[config._attn_implementation]
+        if config.image_pooling_type in {"attention", "attention_meanq"}:
+            self.image_pooling_2d = attention_class(config)
+        elif config.image_pooling_type == "attention_2wide":
+            self.image_pooling_2d = attention_class(config)
+        elif config.image_pooling_type == "attention_v2":
+            self.image_pooling_2d = attention_class(config)
+        elif config.image_pooling_type in [None, "stack"]:
+            self.image_pooling_2d = None
+        else:
```

Comment:
this is overly complicated and:
1. AFAIK, only one model was released so we don't need all the if elses